### PR TITLE
Combine recurring events and add rrule

### PR DESCRIPTION
**Pull request by SteffiStuffel**
### Short description
For the new endpoint /events/?combine_recurring=True only one event object is returned. This object contains an rrule, which defines how the element is repeated.

### Proposed changes
Added this new endpoint, we need to keep the old one for backwards compatibility.
If we want to remove the old endpoint in the future (we should wait at least 6 months), i can open a ticket, i think this would be nice to remove the legacy code.

### Resolved issues

Fixes: #1965 

__________________________________________________

[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

### CODE REVIEW

**Review by JoeyStk — APPROVED** [review]
Thank you very much. Works as expected! Only note would be to update the changelog file :)

**Review by timoludwig — CHANGES_REQUESTED** [review]
Thanks a lot, very cool feature! :tada:

In general, it almost works as expected :+1:

There is just one edge case when the initial occurrence of an recurrent event is already in the past, then we would completely hide this event in the app if we omit the future occurrences in the API.

Also, I had the idea to use the dateutil dependency to generate the rrule strings, even if it's not too complicate to assemble the strings manually.

File: integreat_cms/api/v3/events.py [hunk 1]
```diff
@@ -89,6 +89,9 @@ def transform_event_translation(event_translation, recurrence_date=None):
         "location": transform_poi(event.location),
         "event": transform_event(event, recurrence_date),
         "hash": None,
+        "recurrence_rule": event.recurrence_rule.to_ical_rrule_string()
+        if event.recurrence_rule
+        else None,
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply integreat_cms/api/v3/events.py:94]: I don't think so, event holds the information for one specific event, but recurrence_rule is shared between all events, like e.g. location.

[the remark this answers is not in the document]
Comment (timoludwig) [reply integreat_cms/api/v3/events.py:94]: Ah yes makes sense :+1:

**Comment by SteffiStuffel** [comment]
> 
> There is just one edge case when the initial occurrence of an recurrent event is already in the past, then we would completely hide this event in the app if we omit the future occurrences in the API.
> 

Fix this and added a test for this case.

File: tests/cms/models/events/test_recurrence_rule.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        )
+
+    def test_api_rrule_bimonthly_until(self):
+        recurrence_rule = RecurrenceRule(
+            frequency="MONTHLY",
+            interval=2,
+            weekdays_for_weekly=None,
+            weekday_for_monthly=6,
+            week_for_monthly=1,
+            recurrence_end_date=datetime.datetime(2030, 10, 19),
```
Comment (timoludwig) [inline tests/cms/models/events/test_recurrence_rule.py:92]:
```suggestion
            recurrence_end_date=datetime.date(2030, 10, 19),
```
Since the database field contains a `date` and not a `datetime`, it's maybe better to use this in the test as well (though it doesn't seem to break anything in this case since `datetime.combine()` apparently also works when given a `datetime`)...

File: tests/cms/models/events/test_recurrence_rule.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+from integreat_cms.cms.models import RecurrenceRule, Event
+
+
+# pylint: disable=missing-docstring
+
+
+class TestCreatingIcalRule:
+    """
+    Test whether to_ical_rrule_string function is calculating the rrule corretly
```
Comment (timoludwig) [inline tests/cms/models/events/test_recurrence_rule.py:16]:
```suggestion
    Test whether to_ical_rrule_string function is calculating the rrule correctly
```

File: tests/cms/models/events/test_recurrence_rule.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    """
+
+    test_event = Event(
+        start=datetime.datetime(2030, 1, 1, 11, 30, 0, 0, pytz.UTC),
+        end=datetime.datetime(2030, 1, 1, 12, 30, 0, 0, pytz.UTC),
+    )
+
+    def test_api_rrule_every_year_start_date_in_the_past(self):
```
Comment (timoludwig) [inline tests/cms/models/events/test_recurrence_rule.py:24]: > > There is just one edge case when the initial occurrence of an recurrent event is already in the past, then we would completely hide this event in the app if we omit the future occurrences in the API.
>
> Fix this and added a test for this case.

Actually, this test case doesn't really cover the previous bug (since the real API code is not executed here)... if we really wanted to make sure this isn't reintroduced in the future, we would have to add a past event to the test data and adapt all the expected API responses again... I'll leave this up to you if you're still motivated :sweat_smile:

File: tests/api/expected-outputs/augsburg_ar_events.json [hunk 5]
```diff
@@ -315,6 +320,7 @@
       "recurrence_id": 1,
       "timezone": "Europe/Berlin"
     },
-    "hash": null
+    "hash": null,
+    "recurrence_rule": "DTSTART:20300101T120000\nRRULE:FREQ=WEEKLY;INTERVAL=2;UNTIL=20310101T235959;BYDAY=TU,TH"
   }
-]
+]
```
Comment (timoludwig) [inline tests/api/expected-outputs/augsburg_ar_events.json:326]:
```suggestion
]

```
I'd prefer newlines at the end of files

File: CHANGELOG.md [hunk 6]
```diff
@@ -3,6 +3,7 @@ UNRELEASED
 
 * [ [#2037](https://github.com/digitalfabrik/integreat-cms/issues/2037) ] Add organization and barrier-free fields to POI
 * [ [#2064](https://github.com/digitalfabrik/integreat-cms/issues/2064) ] Show Linkcheck status codes to all users
+* [ [#1965](https://github.com/digitalfabrik/integreat-cms/issues/1965) ] Add ical rrule for recurring events
```
Comment (timoludwig) [inline CHANGELOG.md:6]: Sorry, released a new version yet again :see_no_evil:

**Review by timoludwig — CHANGES_REQUESTED** [review]
Thanks a lot for your changes! :muscle: 

I just have a few more little nitpicky comments, then this is ready to be merged :tada:

File: integreat_cms/api/v3/events.py [hunk 7]
```diff
@@ -136,17 +139,17 @@ def transform_event_recurrences(event_translation, today):
...
 
-    # In order to avoid unnecessary computations, check if any future event
-    # may be valid and return early if that is not the case
-    if (
+    event_is_invalid = (
         recurrence_rule.recurrence_end_date
         and recurrence_rule.recurrence_end_date < today
-    ):
+    )
+    if event_is_invalid:
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply integreat_cms/api/v3/events.py:152]: Leads to unused variable linting error. Will keep it as it is.

File: integreat_cms/api/v3/events.py [hunk 8]
```diff
@@ -185,16 +188,20 @@ def events(request, region_slug, language_slug):
...
     result = []
     now = timezone.now().date()
     for event in region.events.prefetch_public_translations().filter(archived=False):
         event_translation = event.get_public_translation(language_slug)
         if event_translation:
-            if event.end_local.date() >= now:
+            if event.end_local.date() >= now or (
+                event.is_recurring
+                and event.recurrence_rule.to_ical_rrule().after(timezone.now(), True)
+            ):  # or is recurring and last recurring is >= now
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply integreat_cms/api/v3/events.py:200]: I added this comment, but in general i don't think these types of comments are helpful or are any good when thinking about clean code principles.

File: integreat_cms/cms/models/events/recurrence_rule.py [hunk 9]
```diff
@@ -201,6 +205,44 @@ def get_repr(self):
         """
         return f"<RecurrenceRule (id: {self.id}, event: {self.event.best_translation.slug})>"
 
+    def to_ical_rrule(self):
+        """
+        Calculates the ical standardized rrule for a recurring rule. See details of the rrule here:
+        https://icalendar.org/iCalendar-RFC-5545/3-8-5-3-recurrence-rule.html
+
+        :return: The ical rrule for the recurrence rule
+        :rtype: rrule
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply integreat_cms/cms/models/events/recurrence_rule.py:214]: Nice, thank you :rocket:

File: tests/cms/models/events/test_recurrence_rule.py (hunk 4 above)
Comment (SteffiStuffel) [reply tests/cms/models/events/test_recurrence_rule.py:24]: You are right, actually this only shows me, there is too much logic in the endpoints, that should be in the models.

Comment (timoludwig) [reply tests/cms/models/events/test_recurrence_rule.py:23]: Yes, agreed :grin:

File: integreat_cms/api/v3/events.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (timoludwig) [reply integreat_cms/api/v3/events.py]: Ok, for me it would also be fine to remove the comment.
But the previous version with the hanging inline comment after a multi-line if-statement looked less readable to me, this is why I suggested to move the comment into its own line.
Maybe it would be even better to create a new property to the event model which explains what this if-condition is doing? E.g. something along the lines of:

```
@cached_property
def is_past(self):
    """
    This property checks whether an event lies in the past, including potential future recurrences.

    :return: Whether event lies in the past
    :rtype: bool
    """
    now = timezone.now()
    duration = event.end_local - event.start_local
    future_recurrence = (
        self.is_recurring
        and self.recurrence_rule.to_ical_rrule().after(now - duration, True)
    )
    return self.end_local.date() < now.date() and not future_recurrence
```

then, we could use the simpler and more readable property in the API:

```
if not event.is_past:
    result.append(transform_event_translation(event_translation))
```

I also noticed that the `rrule.after()` method only takes the start datetime into account, so in case the last occurrence is currently ongoing, the condition `after(now)` would not return the current instance. So I suggest to calculate the potential start date of the currently ongoing occurrence by subtracting the duration of the event from the current time.
This still has the caveat of an inconsistency (non-recurring events are still shown on their last day even after they're over, and recurring events disappear as soon as the end time has been reached), but I'm not sure which of both behaviors is the desired one?

**Review by timoludwig — APPROVED** [review]
Awesome, thanks! :rocket:
